packages/python/plotly/plotly/tests/test_core/test_update_objects/test_select_subplots.py

```diff
@@ -0,0 +1,182 @@
+        ),
+        (
+            dict(dims=["all", [2, 4, 5]], alls=[all_rows, all_cols], product=True),
+            set([(r, c) for r in all_rows for c in [2, 4, 5]]),
+        ),
+        (
+            dict(dims=["all", 3], alls=[all_rows, all_cols], product=False),
+            set([(all_rows[0], 3)]),
+        ),
+        (
+            dict(dims=["all", 3], alls=[all_rows, all_cols], product=True),
+            set([(r, c) for r in all_rows for c in [3]]),
+        ),
+        (
+            dict(dims=[[1, 3], "all"], alls=[all_rows, all_cols], product=False),
+            set(zip([1, 3], all_cols)),
+        ),
+        (
+            dict(dims=[[1, 3], "all"], alls=[all_rows, all_cols], product=True),
+            set([(r, c) for r in [1, 3] for c in all_cols]),
+        ),
+        (
+            dict(dims=[[1, 3], [2, 4, 5]], alls=[all_rows, all_cols], product=False),
+            set(zip([1, 3], [2, 4, 5])),
+        ),
+        (
+            dict(dims=[[1, 3], [2, 4, 5]], alls=[all_rows, all_cols], product=True),
+            set([(r, c) for r in [1, 3] for c in [2, 4, 5]]),
+        ),
+        (
+            dict(dims=[[1, 3], 3], alls=[all_rows, all_cols], product=False),
+            set([(1, 3)]),
+        ),
+        (
+            dict(dims=[[1, 3], 3], alls=[all_rows, all_cols], product=True),
+            set([(r, c) for r in [1, 3] for c in [3]]),
+        ),
+        (
+            dict(dims=[2, "all"], alls=[all_rows, all_cols], product=False),
+            set([(2, all_cols[0])]),
+        ),
+        (
+            dict(dims=[2, "all"], alls=[all_rows, all_cols], product=True),
+            set([(r, c) for r in [2] for c in all_cols]),
+        ),
+        (
+            dict(dims=[2, [2, 4, 5]], alls=[all_rows, all_cols], product=False),
+            set([(2, 2)]),
+        ),
+        (
+            dict(dims=[2, [2, 4, 5]], alls=[all_rows, all_cols], product=True),
+            set([(r, c) for r in [2] for c in [2, 4, 5]]),
+        ),
+        (dict(dims=[2, 3], alls=[all_rows, all_cols], product=False), set([(2, 3)])),
+        (dict(dims=[2, 3], alls=[all_rows, all_cols], product=True), set([(2, 3)])),
+    ],
+)
+def test_indexing_combinations(test_input, expected):
+    assert set(_indexing_combinations(**test_input)) == expected
+
+
+def _sort_row_col_lists(rows, cols):
+    # makes sure that row and column lists are compared in the same order
+    # sorted on rows
+    si = sorted(range(len(rows)), key=lambda i: rows[i])
+    rows = [rows[i] for i in si]
+    cols = [cols[i] for i in si]
+    return (rows, cols)
+
+
+# _indexing_combinations tests most cases of the following function
+# we just need to test that setting rows or cols to 'all' makes product True,
+# and if not, we can still set product to True.
+@pytest.mark.parametrize(
+    "test_input,expected",
+    [
+        (
+            ("all", [2, 4, 5], False),
+            _unzip_pairs([(r, c) for r in range(1, NROWS + 1) for c in [2, 4, 5]]),
```

Comment:
```suggestion
            zip(*itertools.product(range(1, NROWS + 1), [2, 4, 5]))
```

Reply:
I think the above does the same as your custom `_unzip_pairs` function (you'd have to import `itertools` as well, it's part of the standard library), so if it's possible to reduce the amount of home-written code it's better for maintenance and readability.

Reply:
Nice, I like this trick!